core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -1281,13 +1389,323 @@ public String toStringResponseOnly(Function<? super ResponseHeaders, ?> headersS
         return sanitized != null ? sanitized.toString() : "<sanitized>";
     }
 
-    private static final class ListenerEntry {
-        final RequestLogListener listener;
+    private static final class RequestLogFuture extends EventLoopCheckingFuture<RequestLog> {
+
         final int interestedFlags;
 
-        ListenerEntry(RequestLogListener listener, int interestedFlags) {
-            this.listener = listener;
+        RequestLogFuture(int interestedFlags) {
             this.interestedFlags = interestedFlags;
         }
+
+        void completeLog(RequestLog log) {
+            super.complete(log);
+        }
+
+        @Override
+        public boolean complete(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public void obtrudeValue(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void obtrudeException(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /**
+     * A {@link RequestLog} that delegates to {@link DefaultRequestLog} to read properties without
+     * checking availability.
+     */
+    private final class CompleteRequestLog implements RequestLog {
+
+        @Override
+        public boolean isComplete() {
+            return true;
+        }
+
+        @Override
+        public boolean isRequestComplete() {
+            return true;
+        }
+
+        @Override
+        public boolean isAvailable(RequestLogProperty property) {
+            return true;
```

Comment:
I think we should treat `REQUEST_FIRST_BYTES_TRANSFERRED_TIME` and `RESPONSE_FIRST_BYTES_TRANSFERRED_TIME` specially.

Reply:
We had to previously, but this PR introduces new flags `{request,response}FirstBytesTransferredTimeNanosSet`. When a request is complete, `{REQUEST,RESPONSE}_FIRST_BYTES_TRANSFERRED_TIME` becomes available now. As a result, no properties are treated specially.

Reply:
Oops, I didn't notice that the logic changed. My bad. 😅